src/mongo_odm/query_builder.hpp

```diff
@@ -120,15 +481,53 @@ class NotExpr {
     }
 
    private:
-    const ComparisonExpr<Base, T> _expr;
+    const Expr _expr;
 };
 
+/**
+ * This represents an expression without a field name, for instance used inside $elemMatch queries
+ * for scalar arrays.
+ * The class wraps an existing query expression, and omits the field name when converting to BSON.
+ */
+template <typename Expr>
+class FreeExpr {
+   public:
+    constexpr FreeExpr(Expr expr) : expr(expr) {
+    }
+
+    /**
+     * Appends this expression to a BSON builder, in the form {$op: val} (without a field name)
+     * @param builder A BSON core builder.
+     * @param wrap    Whether to contain this expression inside its own document.
+     */
+    void append_to_bson(bsoncxx::builder::core &builder, bool wrap = false) const {
+        if (wrap) {
+            builder.open_document();
+        }
+        // Append expression without name
+        expr.append_to_bson(builder, false, true);
+
+        if (wrap) {
+            builder.close_document();
+        }
+    }
+
+    // This should never be top-level, and thus does not have a conversion operator directly to
+    // BSON.
+
+   private:
+    const Expr expr;
+};
+
+/**
+ * This represents a list of expressions.
+ */
 template <typename Head, typename Tail>
 class ExpressionList {
    public:
     /**
      * Constructs an expression list.
-     * @param head  The first element in the list
+     * @param head  The first element in the list.
      * @param tail  The remainder of the list
      */
     constexpr ExpressionList(const Head &head, const Tail &tail) : _head(head), _tail(tail) {
```

Comment:
I'd consider re-working this into a non-linked list variant type because, in general, variants and direct expansions are cheaper than recursive variants.  We can push it off after this review, but it can make a big difference when instantiating large numbers of templates.

I.e. something like:

```cpp
template <typename Map, typename... Ts, size_t... idxs>
constexpr void tupleForEachImpl(const std::tuple<Ts...>& tup, Map&& map, std::index_sequence<idxs...>) {
    std::initializer_list<int>{(map(idxs, std::get<idxs>(tup)), 0)...};
}

template <typename Map, typename... Ts>
constexpr void tupleForEach(const std::tuple<Ts...>& tup, Map&& map) {
    return tupleForEachImpl(tup, std::forward<Map>(map), std::index_sequence_for<Ts...>());
}

template <typename... Args>
class ExpressionList {
  ExpressionList(const Args& ...args) : storage(std::make_tuple(args...)) {}

  void append_to_bson(builder::core& builder, bool wrap=false) const {
    tupleForEach(storage, [&](const auto& v) { v.append_to_bson(builder, wrap); });
  }

  std::tuple<Args...> storage;
};
```